source attribute is automatically promoted to sources

illicitonion opened the pull request:
This means that either the `source` or `sources` attribute can be used
for any rule which expects sources. Places that `source` was expected
still verify that the correct number of sources are actually present.
Places that `sources` was expected will automatically promote `source`
to `sources`.

This is a step towards all `sources` attributes being
`EagerFilesetWithSpec`s, which will make them cached in the daemon, and
make them easier to work with with both v2 remote execution and in the
v2 engine in general. It also provides common hooks for input file
validation, rather than relying on them being done ad-hoc in each
`Target` constructor.

For backwards compatibility, both attributes will be populated on
`Target`s, but in the future only the sources value will be provided.

`sources` is guaranteed to be an `EagerFilesetWithSpec` whichever of
these mechanisms is used.

A hook is provided for rules to perform validation on `sources` at build
file parsing time. Hooks are put in place for non-contrib rule types
which currently take a `source` attribute to verify that the correct
number of sources are provided. I imagine at some point we may want to
add a "file type" hook too, so that rules can error if files of the
wrong type were added as sources.

This is a breaking change for rules which use both the `source` and `sources` attributes (and where the latter is not equivalent to the former), or where the `source` attribute is used to refer to something other than a file. `source` is now becoming a
reserved attribute name, as `sources` and `dependencies` already are.

This is also a breaking change for rules which use the `source`
attribute, but never set `sources` in a Payload. These will now fail to
parse.

This is also a slightly breaking change for the `page` rule - before,
omitting the `source` attribute would parse, but fail at runtime. Now,
it will fail to parse.

This is also a breaking change in that in means that the source
attribute is now treated like a glob, and so if a file is specified
which isn't present, it will be ignored instead of error. This feels a
little sketchy, but it turns out we did the exact same thing by making
all sources lists be treated like globs...

illicitonion commented:
@stuhood This is now ready for review. Each commit is standalone reviewable.

I have two open questions:
1. How can I write a unit test along the lines of:
> Given BUILD file with contents X, and files in directory named Y and Z, this error is raised by the engine hydrating the SourcesField

(I want to add general tests for sources hydration which cover these cases: https://github.com/pantsbuild/pants/pull/5908/commits/8c8e6f0c88ffe7a9ed8e6e06367d8676f63e6306 but can't quite work out what the execution request I'm trying to craft is... Ideally this would be a unit test, not an integration test.) I'm specifically trying to write a test which exercises both the extracting _and_ the hydrating of a `SourcesField`.
2. If I revert https://github.com/pantsbuild/pants/pull/5908/commits/3e667b30fae18efdb140977131a6916c736beb99 and run `tests/python/pants_test/backend/docgen/targets:wiki_page` I get:
```
                      E       TargetDefinitionException: Invalid target BuildFileAddress(src/docs/BUILD, readme3): Failed to instantiate Target with type <class 'pants.backend.docgen.targets.doc.Page'>: Subsystem "Arguments" not initialized for scope "target-arguments". Is subsystem missing from subsystem_dependencies() in a task?
```
The reason I'm now hitting this codepath is that the `source` kwarg is now being passed, and the "ignore the source kwarg" codepath is getting hit; as far as I can tell, nothing in the `TestBase.target` codepath initialises the `Target` subsystem, but I don't know how subsystems are meant to be initialised for test. I tried calling `subsystem_util.init_subsystem(Target)` and that didn't help... Any ideas?

illicitonion commented:
Thanks @jsirois! (Turns out I was trying to init in a bad place :))

stuhood commented:
>  How can I write a unit test along the lines of: ... (I want to add general tests for sources hydration which cover these cases: 8c8e6f0 but can't quite work out what the execution request I'm trying to craft is... Ideally this would be a unit test, not an integration test.) I'm specifically trying to write a test which exercises both the extracting and the hydrating of a SourcesField.

There are two unit test APIs for "creating targets" on `python.pants_test.test_base.TestBase`: 1) `self.add_to_build_file` and 2) `self.make_target`. The former will trigger the full pipeline of parsing in the engine and then injection into the graph... the latter will synthetically create a target in memory and then inject it into the graph.

The relationship between these two methods is quite similar to the relationship between what happens when we parse a target from a BUILD file vs what happens when we inject a target during codegen (respectively).

So, I think there are two options:

1. use `self.add_to_build_file` and (optionally) strip away some of the boilerplate around that
2. deprecate or overhaul `self.make_target` to use more of the engine codepath, rather than directly injecting into the BuildGraph.

More generally, the long term goal is to remove `BuildGraph.inject_target` entirely in favor of using `@rule`s and variants/subsystems for codegen and optional target creation, respectively. So it would seem to me that bias'ing toward cleaning up API 1 would be reasonable.

stuhood commented on src/python/pants/engine/legacy/graph.py:
```diff
@@ -386,10 +387,11 @@ def hydrate_sources(sources_field, glob_match_error_behavior):
     sources_field.address.spec_path,
     sources_field.filespecs,
     snapshot)
+  sources_field.validate_fn(fileset_with_spec)
```
If you have appetite for it you could lean more into the design of #4535, and do subclassing of `Field` to allow for additional field types and validations. Someone who wanted to validate "singular" vs "plural" could subclass `Field` to do it. A sticking point for that is allowing for installation of the additional field types so that plugins can do subclassing of `Field`: see the last comment on https://github.com/pantsbuild/pants/issues/4535#issuecomment-380654441 for more context.

illicitonion replied:
I like the idea, but I'm going to punt on this for now :)

stuhood commented on src/python/pants/engine/legacy/graph.py:
```diff
@@ -361,7 +361,8 @@ def hydrate_target(target_adaptor):
 
 
-def _eager_fileset_with_spec(spec_path, filespec, snapshot, include_dirs=False):
+# This is deprecated and should not be used in new code.
```
Did you rename this for the purposes of the unit test? Should be able to leave it private and test it if so.

illicitonion replied:
Undone

stuhood commented on tests/python/pants_test/engine/sources_test_base.py:
```diff
@@ -0,0 +1,25 @@
+
+
+class SourcesTestBase(SchedulerTestBase):
+  def sources_for(self, file_paths):
```
Consider testing `@rule def hydrate_sources` rather than `eager_fileset_with_spec`?

illicitonion replied:
Done

stuhood commented on tests/python/pants_test/engine/sources_test_base.py:
```diff
@@ -0,0 +1,25 @@
+from pants_test.engine.scheduler_test_base import SchedulerTestBase
+
+
+class SourcesTestBase(SchedulerTestBase):
```
Rather than bringing along `SchedulerTestBase`, this should maybe require that it be mixed into `TestBase`? We'll probably want to phase out `SchedulerTestBase` over time...

illicitonion replied:
Moved to `TestBase`

illicitonion commented:
@stuhood With subclassing of fields, would you be expecting a `source` argument to be hydrated to a `AtMostOneSourcesField` which gets hydrated to an `EagerFilesetWithSpec` which is guaranteed to have 0-1 source in it? Where would you be writing that type down in the `Target`/`TargetAdaptor`?

stuhood commented:
All of that needs design.

Currently `TargetAdaptor` defaults to returning a `field_adaptor` for `SourcesField`... which fields and subclasses are legal for which target types is part of what would need working out.

illicitonion commented on src/python/pants/engine/legacy/structs.py:
```diff
@@ -37,7 +38,18 @@ def get_sources(self):
+    source = getattr(self, 'source', None)
     sources = getattr(self, 'sources', None)
+
+    if source and sources:
```
These changes are in `TargetAdaptor`?

stuhood replied:
Oh, derp. Sorry. I think I looked at the filename.

stuhood replied:
I think to be appropriately specific, these should be `if x is not None` checks... it matters less for this one, but might matter below.

illicitonion replied:
Done

illicitonion commented:
Simplified a bunch, and rebased on https://github.com/pantsbuild/pants/pull/5939 (upon which this PR now depends)

stuhood commented on src/python/pants/engine/legacy/structs.py:
```diff
@@ -38,7 +39,18 @@ def get_sources(self):
+        'Cannot specify both source and sources attribute.'
+      )
+
+    if source:
```
This should probably be `if source is not None`, to minimize edgecases around empty strings.

illicitonion replied:
Done

illicitonion commented:
Addressed all comments.

Edited description to note that the `source` attribute will now be treated like a `glob`, because that's how the engine interprets all paths, and that this is technically a breaking change.
